core/base/composition.cpp

```diff
@@ -77,14 +87,26 @@ std::unique_ptr<LinOp> apply_inner_operators(
         // swap in and out
         auto in = std::move(out);
         // build new intermediate vector
-        out_dim[0] = operators[i]->get_size()[0];
+        auto op_size = operators[i]->get_size();
+        out_dim[0] = op_size[0];
         auto out_size = out_dim[0] * num_rhs;
         auto out_data =
             data + (reversed_storage ? storage_size - out_size : size_type{});
         reversed_storage = !reversed_storage;
         out = Dense::create(exec, out_dim,
                             Array<ValueType>::view(exec, out_size, out_data),
                             num_rhs);
+        // for operators with initial guess: set initial guess
+        if (operators[i]->apply_uses_initial_guess()) {
+            if (op_size[0] == op_size[1]) {
+                // square matrix: we can use the previous output
```

Comment:
should we leave an option to use zero as initial guess in square matrix?

Reply:
I would argue that this doesn't really make a difference, since for a few algorithms I can think of (IR, GMRES, CG, ...), the next iteration after `0` would always be `λb` for some scalar λ. I'm not sure we need to be *that* configurable.